pkg/bloomgateway/util.go

```diff
@@ -0,0 +1,129 @@
+package bloomgateway
+
+import (
+	"time"
+
+	"github.com/grafana/loki/pkg/logproto"
+	v1 "github.com/grafana/loki/pkg/storage/bloom/v1"
+	"github.com/prometheus/common/model"
+)
+
+// SliceIterWithIndex implements v1.PeekingIterator
+type SliceIterWithIndex[T any] struct {
+	xs   []T // source slice
+	pos  int // position within the slice
+	idx  int // the index that identifies the iterator
+	zero T   // zero value of T
+}
+
+func (it *SliceIterWithIndex[T]) Next() bool {
+	it.pos++
+	return it.pos < len(it.xs)
+}
+
+func (it *SliceIterWithIndex[T]) Err() error {
+	return nil
+}
+
+func (it *SliceIterWithIndex[T]) At() T {
+	return it.xs[it.pos]
+}
+
+func (it *SliceIterWithIndex[T]) Peek() (T, bool) {
+	if it.pos+1 >= len(it.xs) {
+		return it.zero, false
+	}
+	return it.xs[it.pos+1], true
+}
+
+func (it *SliceIterWithIndex[T]) Index() int {
+	return it.idx
+}
+
+func NewIterWithIndex[T any](i int, xs []T) *SliceIterWithIndex[T] {
+	return &SliceIterWithIndex[T]{
+		xs:  xs,
+		pos: -1,
+		idx: i,
+	}
+}
+
+func getDay(ts model.Time) int64 {
+	return ts.Unix() / int64(24*time.Hour/time.Second)
+}
+
+func getDayTime(ts model.Time) time.Time {
+	return time.Date(ts.Time().Year(), ts.Time().Month(), ts.Time().Day(), 0, 0, 0, 0, time.UTC)
+}
+
+func filterRequestForDay(r *logproto.FilterChunkRefRequest, day time.Time) *logproto.FilterChunkRefRequest {
```

Comment:
Could be the schema's DayTime type:

```suggestion
func filterRequestForDay(r *logproto.FilterChunkRefRequest, day config.DayTime) *logproto.FilterChunkRefRequest {
```